banzai/utils/photometry_utils.py

```diff
@@ -0,0 +1,130 @@
+import requests
+from astropy.table import Table
+from astropy import units
+from astropy.coordinates import SkyCoord
+import numpy as np
+import emcee
+from scipy.optimize import minimize
+from banzai.utils.stats import robust_standard_deviation
+
+
+def get_reference_sources(header, reference_catalog_url):
+    response = requests.post(reference_catalog_url, json=header)
+    response.raise_for_status()
+    return response.json()
+
+
+def match_catalogs(input_catalog, reference_catalog, match_threshold=1.0) -> Table:
+    """
+    Match objects between catalogs by RA and Dec within a threshold (in arcseconds).
+
+    Parameters
+    ----------
+    input_catalog: astropy Table-like
+              Catalog with RA and Dec to be matched
+    reference_catalog: astropy Table-like
+              Reference catalog with RA and Dec to matched
+    match_threshold: float
+                     Threshold in offset to call a source a match in arcseconds
+
+    Returns
+    -------
+    An astropy table with only matched sources but columns from both catalogs
+
+    Notes
+    -----
+    We assume the reference catalog positions are better so we adopt them
+    """
+    input_catalog, reference_catalog = Table(input_catalog), Table(reference_catalog)
+
+    input_coordinates = SkyCoord(ra=input_catalog['ra'], dec=input_catalog['dec'],
+                                 unit=(units.deg, units.deg))
+    reference_coordinates = SkyCoord(ra=reference_catalog['ra'], dec=reference_catalog['dec'],
+                                     unit=(units.deg, units.deg))
+    match_indexes, offsets, _ = input_coordinates.match_to_catalog_sky(reference_coordinates)
+
+    matched_catalog = Table()
+    good_matches = offsets <= match_threshold * units.arcsec
+    for colname in input_catalog.colnames:
+        matched_catalog[colname] = input_catalog[colname][good_matches]
+
+    for colname in reference_catalog.colnames:
+        matched_catalog[colname] = reference_catalog[match_indexes][good_matches]
+
+    return matched_catalog
+
+
+def log_zeropoint_likelihood(theta, mags, mag_errors, catalog_mags, catalog_errors, colors, color_errors):
+    # Our model assumes a zeropoint, color term, and a scatter term
+    # catalog = -2.5 * log10(counts / exptime) + zeropoint + color_term * colors
+    zeropoint, color_term, scatter = theta
+    model = mags + zeropoint + color_term * colors
+    sigma_squared = scatter ** 2.0 + mag_errors ** 2.0 + catalog_errors ** 2.0 + color_errors ** 2.0
+    return -0.5 * np.sum((catalog_mags - model) ** 2.0 / sigma_squared + np.log(2.0 * np.pi * sigma_squared))
+
+
+def fit_photometry(matched_catalog, image_filter, color_to_fit, exptime):
```

Comment:
This may warrant a review from Daniel - I don't feel like I have the domain knowledge to say if this is correct or not. But I am trusting that it is.

Reply:
This is a (very nice, by the way) cleanup of my original fitting code. The approach (likelyhood fitting w/ error estimate) Is great, maybe a bit overkill. But it give you good error estimates, which is definitely a plus.